### DISCUSSION TITLE
Is __declspec(thread) only for windows target?

### DISCUSSION THREAD

**Discussion by sezero** [open-watcom/open-watcom-v2#942] [Q&A]
Is it not possible for os/2 target?

**Comment by jmalak** [comment]
It is not stated explicitly, but in source code is comment that thread-local-storage (TLS) is available for WIN32.
Next in documentation is note that \_\_declspec(thread) is for WIN32.
It looks like it is usable only for WIN32.
I don't know how OS/2 is different from WIN32 for TLS implementation.
In code generator is some stuff for "general TLS" (using run-time functions) but probably it is not finished/active.

**Comment by winspool** [comment]
> Is __declspec(thread) only for windows target? 
> Is it not possible for os/2 target?

Sorry, I don`t know the OpenWatcom code good enough to say something about the
support of thread_local storage in OpenWatcom.

And i have no idea, how thread_local is handled in OS/2.
Do you know of some freely available documentation and/or examples?

This blog entry has a lot of infos/examples about thread local storage, but again, nothing related to OS/2:
https://maskray.me/blog/2021-02-14-all-about-thread-local-storage

Is OS/2 supported by clang/llvm or gcc?
IBM might have tls infos for OS/2 available, but i did not searched their homepage.

---
Since C11, support for thread_local storage is in the C standard: <threads.h>.
For thread_local storage with the key/value semantic,
the C11 API is very similar to the pthreads API.

You can take a look at C11Threads on github.
This project implements many C11 <threads.h> features on top of the pthreads API.
For thread_local storage, only the key/value semantic is supported.
C11Threads includes also an Implementation for Windows,
but again for thread_local features only the parts using the key/value semantic.

C11Threads.h and C11Threads_win.c should help to find the code
in current OpenWatcom related to thread local storage.

---
The most easiest thing for a developer to use thread_local
needs a lot of support from the compiler toolchain and the language runtime: 

The thread_local storage class for a variable.

```
thread_local int example;
thread_local int the_answer = 42;
```

thread_local variables are in a separate memory area
and thread_local variables are accessed in a special way
(on Linux x86/x86_64, relative to the fs segment register)
(See Ulrich Drepper's paper: ELF Handling For Thread-Local Storage)

thread_local variables are collected in an extra section in the elf file.
Values for static initialization of thread_local variables
are in another extra section in the elf file.

And when a new thread is created, memory for the thread_local variables
is reserved in a separate memory block and initialized from the
extra section with the thread_local defaults.

Yes, there are a lot of things going on behind the scenes.
